pandas/algos.pyx

```diff
@@ -1348,6 +1348,56 @@ cdef inline float64_t _median_linear(float64_t* a, int n):
     return result
 
 
+cdef int64_t _checked_add_scalars(int64_t a, int64_t b) except -1:
+    """
+    Performs the addition of two int64 integers but checks
+    that they do not result in overflow first.
+
+    Parameters
+    ----------
+    a : first addend.
+    b : second addend.
+
+    Returns
+    -------
+    sum : The sum of a + b.
+
+    Raises
+    ------
+    OverflowError if a + b exceeds the maximum int64 value.
+    """
+    if np.iinfo(np.int64).max - b < a:
+        raise OverflowError("Python int too large to "
+                            "convert to C long")
+    return a + b
+
+
+def _checked_add_arr_scalar(ndarray[int64_t] arr, int64_t b):
+    """
+    Performs the addition of an int64 array and an int64 integer
+    but checks that they do not result in overflow first.
+
+    Parameters
+    ----------
+    arr : array addend.
+    b : scalar addend.
+
+    Returns
+    -------
+    sum : An array for elements x + b for each element x in arr.
+
+    Raises
+    ------
+    OverflowError if any x + b exceeds the maximum int64 value.
+    """
+    result = []
+
+    for i in arr:
+        result.append(_checked_add_scalars(i, b))
```

Comment:
Appending to a python list like this will be slow, allocate a result array and place values into that (look how other functions in this file do it).